### PR TITLE
17.0 genesis tuto soty

### PR DESCRIPTION
**Pull request by st-yes**

### CODE REVIEW

**Comment by aboo-odoo** [comment]
Ping

**Comment by st-yes** [comment]
@aboo-odoo

**Review by aboo-odoo — COMMENTED** [review]
Good work so far keep going :medal_sports: 

- Update the title to something like [ADD] estate: blablabla and also add a description on the PR
- Try to always add the string parameter to the fields to make it translatable and a bit more readable

_p.s.: I'm being extra picky_

**Comment by aboo-odoo** [comment]
Also, lastly, be mindful of commit titles: [ADD] estate: blablabla

**Comment by st-yes** [comment]
@aboo-odoo

File: estate/models/estate_property.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
-from odoo import models, fields
 from datetime import datetime
 from dateutil.relativedelta import relativedelta
+from odoo import models, fields
```
Comment (aboo-odoo) [inline estate/models/estate_property.py:3]:
```suggestion
from odoo import fields, models
```

File: estate/models/estate_property.py [hunk 2]
```diff
@@ -21,12 +23,13 @@ class EstateProperty(models.Model):
         ('north', 'North'),
         ('south', 'South'),
         ('east', 'East'),
-        ('west', 'West')])
+        ('west', 'West')
```
Comment (aboo-odoo) [inline estate/models/estate_property.py:26]:
```suggestion
        ('west', 'West'),
```

File: estate/views/estate_property_search_view.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+                <field name="expected_price"/>
+                <field name="bedrooms"/>
+                <field name="living_area"/>
+                <field name="facades"/>
+                <!-- filter -->
+                <filter string="Available" name="available" domain="['|', ('state', '=', 'new'), ('state', '=', 'offer_received')]"/>
+                <!-- group by -->
+                <group expand="0" string="Group By">
+                    <filter string="Postcode" name="group_by_postcode" context="{'group_by':'postcode'}"/>
+                </group>
```
Comment (aboo-odoo) [inline estate/views/estate_property_search_view.xml:19]: I guess this is how it's shown in the tutorial (_because the other did it similarly hehe_). But the group is not necessary, the `context="group_by..` is sufficient to tell you're doing a group_by. If you try without, it should do the same
```suggestion
                <filter string="Postcode" name="group_by_postcode" context="{'group_by':'postcode'}"/>
```

**Review by aboo-odoo — COMMENTED** [review]
Well done and thanks for the correction :100: 

Just a few more things:
- Beware of your commit titles. They should be **[TAG] module_name: description**. _STYLE_ is not a valid tag. See [link](https://www.odoo.com/documentation/17.0/contributing/development/git_guidelines.html) for more details.
- Also, don't hesitate to click **Resolve Conversation** when you have done the modifications so that is 'hides' it and I can quickly see how many are left to resolve.

I'll show you next time you're here how rebase works :+1:

**Comment by st-yes** [comment]
@aboo-odoo

File: estate/models/estate_property.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    # relations
+    property_type_id = fields.Many2one('estate.property.type', string="Property Type")
+    buyer_id = fields.Many2one('res.partner', string="Buyer", copy=False)
+    salesperson_id = fields.Many2one('res.users', string="Salesperson")
+    tag_ids = fields.Many2many('estate.property.tag', string="Tags")
+    offer_ids = fields.One2many("estate.property.offer", "property_id", string="Offers")
+
+    # computed field
+    total_area = fields.Integer(compute="_compute_total_area")
+    best_price = fields.Float(compute="_compute_best_price")
```
Comment (aboo-odoo) [inline estate/models/estate_property.py:48]: Nice but there is no need to separate your fields in categories. You can just put them together

File: estate/models/estate_property.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        for record in self:
+            record.total_area = record.garden_area + record.living_area
+
+    @api.depends("offer_ids.price")
+    def _compute_best_price(self):
+        for record in self:
+            if record.offer_ids:
+                record.best_price = max(record.offer_ids.mapped('price'))
+            else:
+                record.best_price = 0
```
Comment (aboo-odoo) [inline estate/models/estate_property.py:62]: We could squash that even more :
```suggestion
            record.best_price = max(record.offer_ids.mapped('price') + [0])
```

File: estate/models/estate_property.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        else:
+            self.garden_area = 0
+            self.garden_orientation = False
+
+    # action methods (assigned to buttons)
+    # error raised if property is canceled
+    def action_set_sold(self):
+        for record in self:
+            if record.state == 'canceled':
+                raise UserError("A canceled property cannot be set as sold.")
```
Comment (aboo-odoo) [inline estate/models/estate_property.py:79]: In Odoo, we try to make all strings translatable. Usually, it is done automatically, like for the fields name. In exceptions, it is not done automatically yet. To set a manual string as translatable, you have to import `_` -> `from odoo import _` and use it as such

```suggestion
                raise UserError(_("A canceled property cannot be set as sold."))
```

File: estate/models/estate_property.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            record.state = 'sold'
+
+    # error raised if property is sold
+    def action_set_canceled(self):
+        for record in self:
+            if record.state == 'sold':
+                raise UserError("A sold property cannot be canceled.")
```
Comment (aboo-odoo) [inline estate/models/estate_property.py:86]: Translation

File: estate/models/estate_property_offer.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        ('accepted', 'Accepted'),
+        ('refused', 'Refused'),
+    ], copy=False)
+
+    # relations
+    partner_id = fields.Many2one('res.partner', string="Partner", required=True)
+    property_id = fields.Many2one('estate.property', string="Property", required=True)
+
+    # computed
+    validity = fields.Integer(default=7)
```
Comment (aboo-odoo) [inline estate/models/estate_property_offer.py:21]: No need to separate here either

File: estate/models/estate_property_offer.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    @api.depends('validity')
+    def _compute_date_deadline(self):
+        for record in self:
+            record.date_deadline = fields.Date.today() + timedelta(days=record.validity)
+
+    def _inverse_date_deadline(self):
+        for record in self:
+            if record.date_deadline:
+                delta = record.date_deadline - fields.Date.today()
+                record.validity = delta.days
```
Comment (aboo-odoo) [inline estate/models/estate_property_offer.py:34]:
```suggestion
                record.validity = (record.date_deadline - fields.Date.today()).days
```

Comment (aboo-odoo) [reply estate/models/estate_property_offer.py:34]: Also, quid if there is no `date_deadline` ?

File: estate/security/ir.model.access.csv [hunk 10]
```diff
@@ -0,0 +1,5 @@
+id,name,model_id/id,group_id/id,perm_read,perm_write,perm_create,perm_unlink
+access_estate_property,access_estate_property,model_estate_property,base.group_user,1,1,1,1
+access_estate_property_type,access.estate.property.type,estate.model_estate_property_type,base.group_user,1,1,1,1
+access_estate_property_tag,access.estate.property.tag,estate.model_estate_property_tag,base.group_user,1,1,1,1
+access_estate_property_offer,access.estate.property.offer,estate.model_estate_property_offer,base.group_user,1,1,1,1
```
Comment (aboo-odoo) [inline estate/security/ir.model.access.csv:5]: Line break

**Pull request closed without merging**